### ISSUE TITLE
Per file ignores?

### ISSUE THREAD

**Issue by memery-imb** [life4/flakehell#34]
Is there a way of specifying linting exceptions on a per-file basis, like with flake8?

e.g.

```
per-file-ignores =
  project/tests/*.py: S101, S105, S404, S603, S607
```

**Comment by orsinium** [comment]
Not yet. We are not sure about the best way to do it in terms of the syntax and merging. Please, tell us about your use-case. Why do you want to exclude particular checks for a particular file?

**Comment by memery-imb** [comment]
The most common exceptions I'd like to make are for tests. flake8-bandit code S101 disallows asserts, which makes sense for a regular function, but isn't compatible with pytest code.

**Comment by orsinium** [comment]
Great! What else? Do you have more cases? What if you would be able to disable some checks only for tests?

**Comment by orsinium** [comment]
Or what if flakehell will have hardcoded exception only for the given checks only for tests? Let's think how limited we can make it for sake of simplicity.

**Comment by sobolevn** [comment]
I thought that `flake8`'s defaults should work, shouldn't they?

**Comment by memery-imb** [comment]
Another example would be in my `datatypes.py` file I have some TypedDicts that require uppercase keys which trips this code: WPS305.

> I thought that flake8's defaults should work, shouldn't they?

I thought so too. It seems like universal ignores work, but per-file ignores don't. I'll try to reproduce.

**Comment by sobolevn** [comment]
I would say that inline `noqa` suits better for this case.

**Comment by memery-imb** [comment]
Hmm, I can't seem to recreate it. It must be a problem on my end. I apologize for wasting your time. I'll close this issue now.

**Comment by orsinium** [comment]
> I'll close this issue now.

Not so fast ;) The main idea behind Flakehell is to group codes by plugins, providing more namespaces and better clarity from where it comes from. We already have plugins with conflicting codes, and this is quite painful to work with. Flakehell solves this problem. Also, Flake8 has really complex algorithms to make a decision about including or excluding checks. Flakehell targeted to solve it as well. I will not disable support for `per-file-ignores` but it's definitely the way to work with codes that we strongly recommend to avoid by all means. So, let's think about all the cases.

> Another example would be in my datatypes.py file I have some TypedDicts that require uppercase

How many you have of them? If less than, let's say, 6, per-line ignore is definitely the way to go.

> The most common exceptions I'd like to make are for tests. flake8-bandit code S101 disallows asserts

Let's be honest, bandit makes no sense for tests at all. Which checks from it could be useful in the context of tests? What if we'll disable `flake8-bandit` for tests at all without a way to configure it?

> I apologize for wasting your time.

No way back, I need all the cases, and it looks like you have them. So, let's talk.

**Comment by memery-imb** [comment]
My datatypes file is filled complicated JSON -> TypedDict conversions so it would be much more than 6!

> Let's be honest, bandit makes no sense for tests at all. Which checks from it could be useful in the context of tests? What if we'll disable flake8-bandit for tests at all without a way to configure it?

I wouldn't have a problem with that.

More per-file-ignores

Here are a few more that apply to tests:

D103: Missing docstring in public function (Usually the test name is descriptive enough)
WPS202: Found too many module members (I usually have one test file per file and don't like limiting my tests)
WPS218: Found too many `assert` statements
WPS118: Found too long name (Paired with the lack of doc-strings)

A few more from datatypes.py
WPS226: Found string constant over-use (I believe I had some issues extracting variable names to constants)
WPS601: Found shadowed class attribute (I think this an artifact of the JSON structure I'm converting)

I also have an orchestrator module that executes many functions in sequence. I ignore this linting error:
WPS201: Found module with too many imports

**Comment by Bahus** [comment]
Here are my 50 cents, for flake 8 and tests I'm using the following configuration:

```ini
per-file-ignores =
    # Exception for unused import to be able to specify module protocol in __init__.py
    __init__.py:F401
    # W0613 – unused argument, since pytest "usefixtures" does not work for fixtures
    # so to specify a dependency we have to pass another fixture as a parameter
    # which won't be used
    # D101 is disabled, since we don't care about docstrings in test classes
    tests/*.py:D101,W0613
```
flakehell completely ignores such configuration and I have to constantly use "noqa".

**Comment by orsinium** [comment]
> Exception for unused import to be able to specify module protocol in __init__.py

When you have an interface in `__init__.py`, specify all imported objects in `__all__` like `__all__ = ['something', 'SomethingElse']`. Then Flake8 will not complain about "imported but unused". PEP-8 recommends to do it as well to keep your interface explicit.

Thank you for list of excluded checks from tests. I think it's a good idea to hardcode them to make flakehell configuration as simple as possible.

> A few more from datatypes.py

Let's think about how to make such cases simpler to handle. I propose to keep such exceptions in the files itself. I guess it would be much more explicit. It depends on when we want to see that something disabled in a file: in the file itself or in the project config. It will definitely benefit if you have a lot of such exceptions, so your config can become quite messy. 

For file it could look like this:

```python
__noqa__ = {
    'wemake-*': ['-WPS226'],
}
```

For config:

```toml
[tool.flakehell.exceptions."project/datatypes.py"]
"wemake-*": ["-WPS226"]
```

Let's choose the best one.

**Comment by Bahus** [comment]
> When you have an interface in __init__.py, specify all imported objects in __all__ like __all__ = ['something', 'SomethingElse']. Then Flake8 will not complain about "imported but unused". PEP-8 recommends to do it as well to keep your interface explicit.

Thank you for pointing this out, I usually pick this approach when developing an application or module from scratch, but in some legacy code (especially if one use asterisks in imports) I feel lazy to fix it :)

I don't think that any part of a tool configuration should be stored in a code, from the first sign is looks useful, however I prefer to store any configuration centralized: easier to read, change, share and maintain. When I decide to switch to another tool I don't need to cleanup all my files – I just need to convert my current configuration. Also... I personally become annoying when there is a lot of different visual "garbage" in a codebase like "noqa: W1010 disable something", "pragma: no cache" and so on.

There is why I consider baseline feature as a configuration of intentionally ignored errors and I don't need to change my code for this.

Long story short: I would store whole flakehell configuration in a pyproject.toml.

**Comment by orsinium** [comment]
Yes, that makes sense. Let's keep per-file settings in the config rather than a file. However, when I'm thinking about asserts and docstrings in tests, I feel like they are should not be checked in the first place, by the plugin itself. The problem is flake8 doesn't provide such an interface for plugins, as I understand. We could make it and rely on it but the goal of lake8 is to make users' life easier with full compatibility in the terms of supported plugins. Even built-in flakehell pylint check can be used with old fashioned flake8 without any troubles. So, I'm still unsure is it good to exclude it in the intermediate level of flakehell itself or not. At least nothing stops to specify the same exceptions explicitly in the config as well. :thinking:

**Comment by memery-imb** [comment]
I would also vote for putting the per-file-ignores in the `pyproject.toml`!

**Comment by webknjaz** [comment]
> I would also vote for putting the per-file-ignores in the `pyproject.toml`!

This would require *also* contributing to flake8 upstream, otherwise, you're basically forcing devs to maintain a duplicate config in two places.

**Comment by orsinium** [comment]
> otherwise, you're basically forcing devs to maintain a duplicate config in two places.

Flakehell compatible with flake8 in some details but not vice versa and not for everything. Our main goal is to make codes configuration simpler and nicer, and `per-file-ignores` in the same category. So, we'll make it only for `pyproject.toml` to keep all plugins configs in one nice place.

**Comment by webknjaz** [comment]
That's sad. I wanted to keep configs where flake8 would understand them and after I'm done with legacy maybe just switch back to flake8.

**Comment by orsinium** [comment]
> I wanted to keep configs where flake8 would understand them

Flake8 logic for codes processing is overcomplicated. Just look at this:

https://github.com/PyCQA/flake8/blob/master/src/flake8/style_guide.py#L271

The main goal of flakehell is to make it simple.

> after I'm done with legacy maybe just switch back to flake8.

What is your motivation to switch back? Flakehell faster and nicer because we can afford to ignore tons of legacy from flake8. Anyway, the goal of the project to make flake8 -> flakehell migration simpler, not vice versa.

**Comment by webknjaz** [comment]
Because my initial impression from the wemake-python-style was that flakehell is mostly there to help with the legacy baseline. But after you mentioned the other goals of this tool I may reconsider.

**Comment by orsinium** [comment]
Moved into https://github.com/wemake-services/wemake-python-styleguide/issues/1097

**Comment by orsinium** [comment]
> flakehell is mostly there to help with the legacy baseline.

Well, mostly, but not the only.
